torch/nn/utils/rnn.py

```diff
@@ -347,3 +347,39 @@ def pack_sequence(sequences, enforce_sorted=True):
     """
     lengths = [v.size(0) for v in sequences]
     return pack_padded_sequence(pad_sequence(sequences), lengths, enforce_sorted=enforce_sorted)
+
+
+def reverse_padded_sequence(inputs, lengths, batch_first=False):
+    """Reverses sequences according to their lengths.
+    
+    `inputs` should have size `T x B x *` if `batch_first` is `False`, or
+    `B x T x *` if `True`. `T` is the length of the longest sequence (or
+    larger), `B` is the batch size, and `*` is any number of dimensions
+    (including 0).
+    
+    Arguments:
+        inputs (Tensor): padded batch of variable length sequences.
+        lengths (list[int]): list of sequence lengths
+        batch_first (bool, optional): if `True`, `inputs` should be `B x T x *`.
+        
+    Returns:
+        A Tensor with the same size as `inputs`, but with each sequence
+        reversed according to its length.
+    """
+    if batch_first:
+        inputs = inputs.transpose(0, 1)
+    max_length, batch_size = inputs.size(0), inputs.size(1)
+    if len(lengths) != batch_size:
+        raise ValueError('inputs is incompatible with lengths.')
+    ind = [list(reversed(range(0, length))) + list(range(length, max_length))
+           for length in lengths]
```

Comment:
@rdipietro I saw the issue you raised, #1794. I didn't read it carefully enough to realise that you had an up-to-date version of [your code which was linked linked there](https://github.com/rdipietro/pytorch/blob/4c00324affb8c6d53d4362e321ea0e99ede6cfde/torch/nn/utils/rnn.py), so I wrote my own version. Although it's slower on the CPU, it's ~4 times faster when using a benchmark similar to @jihunchoi's. i.e.:

```python
rnn = torch.nn.RNN(300, 100).cuda()
t = torch.randn(128,100,300).cuda()
l = torch.arange(100).cuda()
print(timeit.timeit(lambda: rnn(reverse_padded_sequence(t, l)), number=100)) # Mine: 0.40336035192012787
print(timeit.timeit(lambda: rnn(reverse_padded_sequence_rd(t, l)), number=100)) # Yours: 1.6065499729011208
```

I haven't done an "ablation study" on the code to prove why it's faster, but I'm fairly confident that it has to do with how `ind` is being computed in pure Python. The time complexity is _O(max\_length * batch\_size)_, which is significant, so it's better to do it in PyTorch.

Another thing which has the _potential_ to slow it down (but I don't think it did in my benchmark) is this part, a few lines down:

```python
    for dim in range(2, inputs.dim()):
        ind = ind.unsqueeze(dim)
```

The issue here is that the size of the graph will be linear in `inputs.dim()`. Ideally, it'd be constant. You can achieve that with `torch.view` or `torch.reshape` by writing something like `ind.view(ind.shape + (1,) * (inputs.dim() - 2))`.

You can compare this PR to [what I was going to raise](https://github.com/pytorch/pytorch/commit/1b65ed072aa6908ee1266bd1e21275aa0b742a3e) before realising that you raised this. In addition to the performance difference, there's one behavioural difference, which is that the trailing parts of the reversed sequences are masked instead of being "rotated". I can't see that making any difference in practice unless someone forgets to mask their output of their RNN themselves.

Reply:
That being said, having this PR merged would be really valuable to me, with or without the extra performance tweaks. I'd really like to see it merged soon. (It's been opened for two years!) I think I'll wait for your (@rdipietro) reply to see if you want to make these changes, then I'll start making a ruckus to have PyTorch's maintainers merge your PR :smile: